### ISSUE TITLE
[HOLD for payment 2024-08-14] [$500][P2P Distance] Create UpdateMoneyRequestDistanceRate in App

### ISSUE THREAD

**Issue by neil-marcellini** [Expensify/App#36987]
Create UpdateMoneyRequestRate following [this plan](https://docs.google.com/document/d/15GQhipcqOJ69MbFA54KRUcWxjICURHYcfTlAb3eaVEQ/edit#bookmark=id.th2wu72baboy). Please be sure to read and understand the relevant sections of the entire plan as well.
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fde353d38f5fdf1b</li>
        <li>Upwork Job ID: 1760136022052638720</li>
        <li>Last Price Increase: 2024-02-21</li>
<li>Automatic offers: </li>
<ul>
<li>paultsimura | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @stephanieelliott</details>

**Comment by neil-marcellini** [comment]
@paultsimura's proposal looks good. Sounds like we will hold on #36985, so updating the title for that.

**Comment by neil-marcellini** [comment]
Still on HOLD

**Comment by stephanieelliott** [comment]
Still held on https://github.com/Expensify/App/issues/36985,

**Comment by stephanieelliott** [comment]
Still held, but there has been recent progress on https://github.com/Expensify/App/issues/36985,

**Comment by stephanieelliott** [comment]
Still held on https://github.com/Expensify/App/issues/36985, EOW ETA for PR

**Comment by stephanieelliott** [comment]
Still held, blocking PR us under active review

**Comment by neil-marcellini** [comment]
Hey @paultsimura what do you think about getting started on this soon, branching off of the PR that's currently in progress [[App] Feat/36985 create new rate field](https://github.com/Expensify/App/pull/38543)?

**Comment by neil-marcellini** [comment]
> 1. Since for P2P distance requests we are using the default const rate values, the "edit rate" functionality should be enabled only for workspace distance requests?

Yes



> 2\. Is it OK to use the field name simply `rate` (e.g. in the editing route: `/edit/request/rate/{id}` and in const: `EDIT_REQUEST_FIELD.RATE`), or should we lean to specifying that it's a `mileageRate`?

I prefer mileageRate or distanceRate because we have other rates, such as tax.

> 3\. What should the `MODIFIEDEXPENSE` say in case of the modified rate (considering a rate can have a custom name & the rate itself)?

To make this easy for me to answer, here is the backend logic in PHP. I'm sure it's ok to share since  we should probably implement the same thing in App for the optimistic data.

<details>
<summary>Backed logic for distance modified expense message</summary>

```php
    private function getNewDotCommentForDistanceRequest(string $newMerchant, string $oldMerchant, string $newAmount, string $oldAmount): string
    {
        // The format of the merchant in a distance request from NewDot comes in the format:
        // <quantity> <unit> @ <formatted rate with currency> / <unit>
        // Example: 1 mile @ $5 / mi
        // For more examples, see ModifiedExpenseTest.php's testGetNewDotComment
        $distanceMerchantRegex = '/^[0-9.]+ \w+ @ (-|-\()?(\p{Sc}|\w){1,3} ?[0-9.]+\)? \/ \w+$/';
        $changedField = 'distance';

        // If the new & old merchants match the new distance request format, check if the distance or rate changed.
        // Otherwise, just display that the distance changed.
        if (preg_match($distanceMerchantRegex, $newMerchant) && preg_match($distanceMerchantRegex, $oldMerchant)) {
            // Extract distance and rate from merchant. These values are separated by ' @ '.
            $oldValues = explode('@', $oldMerchant);
            $oldDistance = trim($oldValues[0] ?? '');
            $oldRate = trim($oldValues[1] ?? '');
            $newValues = explode('@', $newMerchant);
            $newDistance = trim($newValues[0] ?? '');
            $newRate = trim($newValues[1] ?? '');

            // Determine if this action is for a rate modification.
            if ($oldDistance === $newDistance && $oldRate !== $newRate) {
                $changedField = 'rate';
            }
        } else {
            Log::hmmm("Distance request merchant doesn't match NewDot format. Defaulting to showing as distance changed.", ['newMerchant' => $newMerchant, 'oldMerchant' => $oldMerchant]);
        }
        if (!strlen($oldMerchant)) {
            return "set the $changedField to $newMerchant, which set the amount to $newAmount";
        }
        return "changed the $changedField to $newMerchant (previously $oldMerchant), which updated the amount to $newAmount (previously $oldAmount)";
    }
```
</details>

**Comment by stephanieelliott** [comment]
PR is under review

**Comment by neil-marcellini** [comment]
@paultsimura the new rate field PR has been merged. Agata is working on some follow ups, but that should be done soon [[CP Staging] Feat/36985 create new rate field followups](https://github.com/Expensify/App/pull/40711#top). What's your ETA to raise the PR for this? Would you please link it here as soon as it's ready? I usually like to review before the C+ approves.

**Comment by stephanieelliott** [comment]
Reapplying the `New Feature` label to get another BZ member on this while I am OOO til May 2. Thanks @miljakljajic -- to catch you up on where we are, we are waiting for @paultsimura to raise a PR.  I'll grab this back from you when I return!

**Comment by neil-marcellini** [comment]
Ok yeah good point. I forgot how slowly that's coming along on the backend 😅. I'll let you know when it's ready, and I'll put it on HOLD for now.

**Comment by neil-marcellini** [comment]
On hold still

**Comment by stephanieelliott** [comment]
I'm back from OOO, thanks for watching this while I was out @miljakljajic!

**Comment by stephanieelliott** [comment]
Still held on #36969

**Comment by stephanieelliott** [comment]
Still held on https://github.com/Expensify/App/issues/36969, the PR blocking that one is actively being worked on.

**Comment by neil-marcellini** [comment]
Yep, not overdue

**Comment by stephanieelliott** [comment]
Still holding on https://github.com/Expensify/App/issues/36969

**Comment by neil-marcellini** [comment]
Still holding

**Comment by stephanieelliott** [comment]
Still held on https://github.com/Expensify/App/issues/36969, there is progress happening on that issue though.

**Comment by neil-marcellini** [comment]
This should be off hold in the next few days. The last backend PR was just merged and will deploy to staging tomorrow. @paultsimura would you please get started now or tomorrow?

**Comment by stephanieelliott** [comment]
Applying the New Feature label to get another BZ member on this while I am OOO til July 10. To catch you up on where we are @johncschuster: we're waiting for a PR from @paultsimura. 

Thanks for watching over this, I'll grab it back from you when I return!

**Comment by stephanieelliott** [comment]
Hey @paultsimura does that mean we should put this issue on hold for https://github.com/Expensify/App/issues/36969? Or were you waiting on just one of the linked PRs there (there are a few)

**Comment by stephanieelliott** [comment]
Sounds good -- I'll keep an eye on the issue and wait a bit before marking this GH as HOLD

**Comment by stephanieelliott** [comment]
Hey @paultsimura looks like https://github.com/Expensify/App/issues/36969 was merged to prod, so we should be unblocked here now!

**Comment by stephanieelliott** [comment]
PR is awaiting your review @mananjadhav

**Comment by mananjadhav** [comment]
Thanks for the bump @stephanieelliott. I'll finish this today.

**Comment by stephanieelliott** [comment]
PR was merged to Main earlier today!

**Comment by neil-marcellini** [comment]
@stephanieelliott will you please handle payment and then we can close this out?

**Comment by stephanieelliott** [comment]
Summarizing payment on this issue: 

 - Contributor: @paultsimura $500 via Upwork -- PAID 
 - Contributor+: @mananjadhav  $500 via NewDot -- please request 

@manjesh-yadav @paultsimura as the last step on this issue, can you please propose a regression test? 
 
 Upwork job is here: https://www.upwork.com/jobs/~01fde353d38f5fdf1b

**Comment by mananjadhav** [comment]
Thanks @stephanieelliott.   This is more of a feature request so we don't have any offending PR.

**Regression Test proposal.**

1. Open the app and have a workspace with multiple distance rates.
2. Create a Distance request to the workspace
3. Navigate to the request details page
4. Click the "Rate" field and change the rate
5. Verify the rate & amount are changed correctly
6. Verify the "modified expense" action was added 
7.  Now, create a Distance request to self via "track expense"
8.  Navigate to the request details page
9. Verify the "Rate" field is not clickable

**Comment by stephanieelliott** [comment]
Thanks @mananjadhav! Regression test issue created here: https://github.com/Expensify/Expensify/issues/421162

**Comment by JmillsExpensify** [comment]
$500 approved for @mananjadhav
